src/mon/NVMeofGwMon.cc

```diff
@@ -0,0 +1,524 @@
+          try {
+        return preprocess_command(op);
+          } catch (const bad_cmd_get& e) {
+          bufferlist bl;
+          mon.reply_command(op, -EINVAL, e.what(), bl, get_last_committed());
+          return true;
+        }
+
+        default:
+          mon.no_reply(op);
+          derr << "Unhandled message type " << m->get_type() << dendl;
+          return true;
+      }
+    return false;
+}
+
+bool NVMeofGwMon::prepare_update(MonOpRequestRef op){
+    auto m = op->get_req<PaxosServiceMessage>();
+      switch (m->get_type()) {
+        case MSG_MNVMEOF_GW_BEACON:
+          return prepare_beacon(op);
+
+        case MSG_MON_COMMAND:
+          try {
+        return prepare_command(op);
+          } catch (const bad_cmd_get& e) {
+        bufferlist bl;
+        mon.reply_command(op, -EINVAL, e.what(), bl, get_last_committed());
+        return false; /* nothing to propose! */
+          }
+
+        default:
+          mon.no_reply(op);
+          dout(1) << "Unhandled message type " << m->get_type() << dendl;
+          return false; /* nothing to propose! */
+      }
+    return true;
+}
+
+bool NVMeofGwMon::preprocess_command(MonOpRequestRef op)
+{
+    dout(10) << dendl;
+    auto m = op->get_req<MMonCommand>();
+    std::stringstream ss;
+    bufferlist rdata;
+
+    cmdmap_t cmdmap;
+    if (!cmdmap_from_json(m->cmd, &cmdmap, ss))
+    {
+        string rs = ss.str();
+        dout(1) << "Invalid command "  << m->cmd << "Error " << rs << dendl;
+        mon.reply_command(op, -EINVAL, rs, rdata, get_last_committed());
+        return true;
+    }
+
+    string prefix;
+    cmd_getval(cmdmap, "prefix", prefix);
+    dout(10) << "MonCommand : "<< prefix <<  dendl;
+
+    return false;
+}
+
+bool NVMeofGwMon::prepare_command(MonOpRequestRef op)
+{
+    dout(10)  << dendl;
+    auto m = op->get_req<MMonCommand>();
+    int rc;
+    std::stringstream ss;
+    bufferlist rdata;
+    string rs;
+    int err = 0;
+    cmdmap_t cmdmap;
+    bool response = false;
+
+    if (!cmdmap_from_json(m->cmd, &cmdmap, ss))
+    {
+        string rs = ss.str();
+        mon.reply_command(op, -EINVAL, rs, rdata, get_last_committed());
+        return true;
+    }
+
+    string format = cmd_getval_or<string>(cmdmap, "format", "plain");
+    boost::scoped_ptr<Formatter> f(Formatter::create(format));
+
+    const auto prefix = cmd_getval_or<string>(cmdmap, "prefix", string{});
+
+    dout(10) << "MonCommand : "<< prefix <<  dendl;
+    if( prefix == "nvme-gw create" || prefix == "nvme-gw delete" ) {
+        std::string id, pool, group;
+
+        cmd_getval(cmdmap, "id", id);
+        cmd_getval(cmdmap, "pool", pool);
+        cmd_getval(cmdmap, "group", group);
+        auto group_key = std::make_pair(pool, group);
+        dout(10) << " id "<< id <<" pool "<< pool << " group "<< group << dendl;
+        if(prefix == "nvme-gw create"){
+            rc = pending_map.cfg_add_gw(id, group_key);
+            ceph_assert(rc!= -EINVAL);
```

Comment:
I just noticed this from another change: Why do we assert on EINVAL here? If there's some kind of error shouldn't we just return that to the user?

Reply:
@gregsfortytwo ,This is an indication of the critical error . The internal Database of NVMeofGwMon is corrupted if this error hit . Monitor cannot continue and need to exit,

Reply:
This has been resolved by returning error codes to the invalid callers.